modules/geo-layers/src/imagery-layer/imagery-layer.ts

```diff
@@ -0,0 +1,205 @@
+ */
+export class ImageryLayer extends CompositeLayer<ImageryLayerProps> {
+  static layerName = 'ImageryLayer';
+  static defaultProps: DefaultProps<ImageryLayerProps> = defaultProps;
+
+  /** Lets deck.gl know that we want viewport change events */
+  /*override*/ shouldUpdateState(): boolean {
+    return true;
+  }
+
+  /*override*/ initializeState(): void {
+  }
+  
+  /*override*/ updateState({changeFlags, props, oldProps}: UpdateParameters<this>): void {
+    const state = this.state as ImageryLayerState;
+
+    if (changeFlags.propsChanged) {
+      const dataChanged =
+        changeFlags.dataChanged ||
+        props.serviceType !== oldProps.serviceType || 
+        (changeFlags.updateTriggersChanged &&
+          (changeFlags.updateTriggersChanged.all || changeFlags.updateTriggersChanged));
+
+
+      // Check if data source has changed
+      if (dataChanged) {
+        state.imageSource = createImageSource(this.props);
+        this._loadMetadata();
+        debounce(() => this.loadImage('image source changed'), 0);
+      }
+
+      // Some sublayer props may have changed
+    }
+    
+    if (changeFlags.viewportChanged) {
+      debounce(() => this.loadImage('viewport changed'));
+    }
+
+    const propsChanged = changeFlags.propsOrDataChanged || changeFlags.updateTriggersChanged;
+  }
+  
+  /*override*/ finalizeState(): void {
+    // TODO - we could cancel outstanding requests
+  }
+  
+  /*override*/ renderLayers(): Layer {
+    // TODO - which bitmap layer is rendered should depend on the current viewport
+    // Currently Studio only uses one viewport
+    const state = this.state as ImageryLayerState;
+    const {imageSource} = this.state;
+    const {bounds, image, width, height} = this.state;
+
+    console.log('render', this.state);
+
+    return image && new BitmapLayer({
+      ...this.getSubLayerProps({id: 'bitmap'}),
+      bounds,
+      image,
+      pickable: true, // TODO inherited?
+      onHover: (info) => console.log('hover in bitmap layer', info),
+      onClick: (info) => {
+        // @ts-expect-error
+        const {bitmap, layer} = info;
+        console.log('click in bitmap layer', info);
+        if (bitmap) {
+          const x = bitmap.pixel[0];
+          const y = bitmap.pixel[1];
+          debounce(async () => {
+            const featureInfo = await imageSource.getFeatureInfo({
+              layers: this.props.layers,
+              // todo image width may get out of sync with viewport width
+              width,
+              height,
+              bbox: bounds,
+              query_layers: this.props.layers,
+              x,
+              y,
+              info_format: 'text/plain'
+            })
+            // console.log(featureInfo);
+          }, 0);
+        }
+      }
+    });
+  }
+
+  /** Run a getMetadata on the image service */
+  async _loadMetadata(): Promise<void> {
+    this.props.onMetadataLoadStart();
+    const state = this.state as ImageryLayerState;
```

Comment:
You can type `this.state`, see https://github.com/visgl/deck.gl/blob/8.8-release/modules/layers/src/icon-layer/icon-layer.ts#L141

Reply:
this.state typed